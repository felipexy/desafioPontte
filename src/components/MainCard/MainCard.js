import React from 'react';
import './MainCard.scss';
import { useSelector } from 'react-redux';
import CardTitle from '../CardTitle/CardTitle';
import Slider from '../Slider/Slider';
import CardMonths from '../CardMonths/CardMonths';
import ContinueBtt from '../Buttons/ContinueBtt';
import '../../Utils/loader.scss';
import '../../Icons/questMark.scss';
import Unavailable from '../Unavailable/Unavailable';
import formatValues from '../../Helpers/formatBRL';

const MainCard = () => {
  const sliderPosition = useSelector((state) => state.id);
  const OptChosen = useSelector((state) => state.valid);
  const loading = useSelector((state) => state.loading);
  const options = useSelector((state) => state.data);

  const cardOptions = [];
  let qtdeOptions = 0;

  function defineQtdeOptions() {
    options.prazos.forEach((prazo) => {
      qtdeOptions++;
    });
    return qtdeOptions;
  }

  function optionCardMount(valor) {
    for (let i = 0; i < qtdeOptions; i++) {
      cardOptions.push(
        <CardMonths
          key={i}
          months={options.prazos[i]}
          value={formatValues(options.parcelas[i][valor])}
        />
      );
    }
  }

  if (loading) {
    defineQtdeOptions();
    const valueTitle = options.valoresEmprestimo[sliderPosition];
    const grossAmount = options.valoresEmprestimeBruto[sliderPosition];
    if (qtdeOptions >= 1 && qtdeOptions < 6) {
      return (
        <div className="mainCard">
          <CardTitle />
          <div>
            <h1 data-testid="vlTitle" className="mainCard__valueText">
              R$ {formatValues(valueTitle)}
            </h1>
            <h3 className="mainCard__grossAmount">
              Valor bruto: R$ {formatValues(grossAmount)}{' '}
              <i className="questMarkIcon" />
            </h3>
          </div>
          <div>
            <Slider
              min="0"
              maxPrazos={qtdeOptions - 1}
              minValue={formatValues(options.valoresEmprestimo[0])}
              maxValue={formatValues(
                options.valoresEmprestimo[qtdeOptions - 1]
              )}
            />
            <div>
              <h3 className="mainCard__selectQty">
                Selecione a quantidade de parcelas
              </h3>
            </div>
            <div className="cards">
              {optionCardMount(sliderPosition)}
              {cardOptions}
            </div>
            <ContinueBtt isValid={OptChosen} />
          </div>
          <p className="mainCard__footer">
            Taxa de 1,09% ao mês. Valor da primeira parcela - Sistema de
            Amortização Constante (suas parcelas diminuem com o tempo).
          </p>
        </div>
      );
    }
    return <Unavailable />;
  }
  return <div className="loader" />;
};

export default MainCard;
