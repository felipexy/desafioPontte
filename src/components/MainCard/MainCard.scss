.mainCard {
  display: flex;
  flex-direction: column;
  border-radius: 15px;
  position: relative;
  width: 90%;
  background: white;
  margin-right: auto;
  margin-left: auto;
  top: 100px;
}
  
.mainCard::after {
  content: "";
  display: block;
  padding-bottom: 2%;
}

.MainContent {
  justify-content: center;
  width: fit-content;
  margin-right: auto;
  margin-left: auto;
  color: black;
  height: min-content;
  font-size: 2vw;
}

.mainCard__valueText {
  margin-top: 2%;
  justify-content: center;
  position: relative;
  width: fit-content;
  margin-right: auto;
  margin-left: auto;
  color: rgb(3, 3, 58);
  height: min-content;
  font-size: 5vw;
}

.mainCard__grossAmount {
  margin-top: -4%;
  justify-content: center;
  position: relative;
  width: fit-content;
  margin-right: auto;
  margin-left: 35%;
  color: rgb(3, 3, 58);
  height: min-content;
  font-size: 2vw;
}

.mainCard__selectQty{
  margin-top: 0.4%;
  justify-content: center;
  position: relative;
  width: fit-content;
  margin-right: auto;
  margin-left: auto;
  color: rgb(149, 37, 147);
  height: min-content;
  font-size: 1.9vw;
}

.mainCard__footer{
  margin-right: auto;
  margin-left: auto;
  font-size: 1.2vw;
  font-weight: 500;
  color:  rgb(3, 3, 58);
}
