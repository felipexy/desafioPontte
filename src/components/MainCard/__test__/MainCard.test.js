import React, { useState, useEffect } from 'react';
import ReactDOM from 'react-dom';
import MainCard from '../MainCard';
import { render, cleanup, waitFor, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import { createStore } from 'redux';
import { Provider, useDispatch } from 'react-redux';

afterEach(cleanup);

const apiData = `{
    "prazos": [
        60,
        90,
        120,
        150,
        180
    ],
    "valoresEmprestimo": [
        70000,
        85000,
        100000,
        115000,
        130000
    ],
    "valoresEmprestimeBruto": [
        78919.63,
        95114.83,
        110841.81,
        126366.55,
        141891.29  
    ],
    "parcelas": [
        [
            2287.12,
            2736.97,
            3173.81,
            3605.04,
            4036.26
        ],
        [
            1848.68,
            2208.55,
            2558.02,
            2903,
            3247.98
        ],
        [
            1629.46,
            1944.35,
            2250.13,
            2551.98,
            2853.83 
        ],
        [
            1497.93,
            1785.82,
            2065.39,
            2341.37,
            2617.35
        ],
        [
            1410.24,
            1680.14,
            1942.24,
            2200.96,
            2459.69
        ]
    ],
    "loading": true
}`;

const startingState = {
  id: 2,
  valid: false,
  data: apiData,
  loading: apiData.loading,
};

function reducer(state = startingState, action) {
  switch (action.type) {
    case 'CHANGE_ID':
      return {
        id: action.id,
        valid: state.valid,
        data: state.data,
        loading: state.loading,
      };
    case 'CHANGE_VALID':
      return {
        id: state.id,
        valid: action.valid,
        data: state.data,
        loading: state.loading,
      };
    case 'LOAD_DATA':
      return {
        id: state.id,
        valid: state.valid,
        data: action.data,
        loading: action.loading,
      };
    default:
      return state;
  }
}

function renderWithRedux(
  component,
  { initialState, store = createStore(reducer, initialState) } = {}
) {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
  };
}

describe('Tests for MainCard component with Redux', () => {
  it('Renders with Redux', () => {
    const { getByTestId } = renderWithRedux(<MainCard />);
  });

  /* it("Check if defineQtdeOptions is called", async() => {
        const {getByTestId} = renderWithRedux(<MainCard></MainCard>);
        
        expect(getByTestId( 'vlTitle' )).toHaveTextContent("R$ 100.000,00");
    }) */

  it('Matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={createStore(reducer, startingState)}>
          <MainCard />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
