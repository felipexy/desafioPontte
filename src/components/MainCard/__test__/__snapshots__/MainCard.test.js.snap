// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for MainCard component with Redux Matches snapshot 1`] = `
<div
  className="loader"
/>
`;
